pkg/s3select/input.go

```diff
@@ -0,0 +1,356 @@
+func (reader *Input) createProgressXML() (string, error) {
+	if reader.options.HasHeader {
+		reader.stats.BytesProcessed += processSize(reader.header)
+	}
+	if !(reader.options.Compressed != "") {
+		reader.stats.BytesScanned = reader.stats.BytesProcessed
+	}
+	progressXML := &progress{
+		BytesScanned:   reader.stats.BytesScanned,
+		BytesProcessed: reader.stats.BytesProcessed,
+		BytesReturned:  reader.stats.BytesReturned,
+	}
+	out, err := xml.Marshal(progressXML)
+	if err != nil {
+		return "", err
+	}
+	return xml.Header + string(out), nil
+}
+
+// Header returns the header of the reader. Either the first row if a header
+// set in the options, or c#, where # is the column number, starting with 0.
+func (reader *Input) Header() []string {
+	return reader.header
+}
+
+// Row is a Struct for keeping track of key aspects of a row.
+type Row struct {
+	record string
+	err    error
+}
+
+// Execute is the function where all the blocking occurs, It writes to the HTTP
+// response writer in a streaming fashion so that the client can actively use
+// the results before the query is finally finished executing. The
+func (reader *Input) Execute(writer http.ResponseWriter) error {
+	myRow := make(chan *Row)
+	curBuf.Reset()
+	progressTicker = time.NewTicker(progressTime)
+	continuationTimer = time.NewTimer(continuationTime)
+	defer progressTicker.Stop()
+	defer continuationTimer.Stop()
+	go reader.runSelectParser(convertMySQL(reader.options.Expression), myRow)
+	for {
+		select {
+		case row, ok := <-myRow:
+			if ok && row.err != nil {
+				errorMessage := reader.writeErrorMessage(row.err, curBuf)
+				_, err := errorMessage.WriteTo(writer)
+				if err != nil {
+					return err
+				}
+				curBuf.Reset()
+				close(myRow)
+				return nil
+			} else if ok && len(row.record) > 0 {
+				message := reader.writeRecordMessage(row.record, curBuf)
+				_, err := message.WriteTo(writer)
+				if err != nil {
+					return err
+				}
+				curBuf.Reset()
+				reader.stats.BytesReturned += int64(len(row.record))
+				if !continuationTimer.Stop() {
+					<-continuationTimer.C
+				}
+				continuationTimer.Reset(continuationTime)
+			} else if !ok {
+				statPayload, err := reader.createStatXML()
+				if err != nil {
+					return err
+				}
+				statMessage := reader.writeStatMessage(statPayload, curBuf)
+				_, err = statMessage.WriteTo(writer)
+				if err != nil {
+					return err
+				}
+				curBuf.Reset()
+				message := reader.writeEndMessage(curBuf)
+				_, err = message.WriteTo(writer)
+				if err != nil {
+					return err
+				}
+				return nil
+			}
+
+		case <-progressTicker.C:
+			progressPayload, err := reader.createProgressXML()
+			if err != nil {
+				return err
+			}
+			progressMessage := reader.writeProgressMessage(progressPayload, curBuf)
+			_, err = progressMessage.WriteTo(writer)
+			writer.(http.Flusher).Flush()
```

Comment:
Once we change the function signature lets look for `writer.(http.Flusher)` first like

```
flusher, ok := writer.(http.Flusher)
if ok {
   flusher.Flush()
}
```